---
title: "ProcessingandAnlyzingFullModel"
author: "Alison Paquette"
date: "9/29/2020"
output: html_document
---

* NOTE: Overall this is very sequential. I thought I could fix it but due to the way that igraph wants the data it is much easier to work withint he cytoscape style 
# Load Packages
```{r}
library(igraph)
library(RColorBrewer)
```


# Load all Relevant Intermediate Data
```{r}
# The easy to access compiled model
load("~/PlacentalTRN/IntermediateData/GenomeScaleModel6242020.RData")
rownames(Model)<-paste0(Model$targetGene,"_",Model$TF)

# LASSO Regression Values from Holdout Dataset
load("~/PlacentalTRN/IntermediateData/HoldoutData_LASSO.RData")
#Interaction Terms
load("~/PlacentalTRN/IntermediateData/FetalSexInteractionTerms.RData")
Interactions_Sig<-subset(Interactions,Q_interaction<0.05)
rownames(Interactions_Sig)<-paste0(Interactions_Sig$Target_Gene,"_",Interactions_Sig$TF)



# List of Placental Genes from Human Protein Atlas
HPA_Genes<-read.csv("~/RandomImpFiles/UpdatedHPAList.csv")
HPA_Genes<-as.character(HPA_Genes$Gene) #494 Genes considered "elevated" in the placenta as compared to other tissues https://www.proteinatlas.org/humanproteome/tissue/placenta


```

#Fix Holdout LASSO To contain only what we need
```{r}
TG_Table<-merge(TG_Table,Holdout_LASSO,by.x="targetGene",by.y="TG",all=F)
TG_Table$RSquared<-as.numeric(paste0(TG_Table$RSquared))
```

# Make Cytoscape Files

```{r}
Edge<-Model

rownames(Edge)<-paste0(Edge$targetGene,"_",Edge$TF)

Edge$Dir<-as.numeric(Edge$Cor)
Edge$Dir[Edge$Dir>0] <- "POS"
Edge$Dir[Edge$Dir<=0] <- "NEG"
table(Edge$Dir)

#Make it pretty for Cytoscape
Edge$Strength<-round(abs(Edge$Cor),2)
Edge$Cor<-round((Edge$Cor),2)


#Node File
TFs<-cbind(unique(Edge$TF),rep("TF",length(unique(Edge$TF))))
TGs<-cbind(unique(Edge$targetGene),rep("TargetGene",length(unique(Edge$targetGene))))


TGs_TF<-intersect(unique(Edge$targetGene),unique(Edge$TF)) #344 TFs which are also target genes
TGs<-TGs[-(match(TGs_TF,TGs[,1])),]

Node<-as.data.frame(rbind(TFs,TGs))

colnames(Node)<-c("Gene","Type")
table(Node$Type)

Node<-merge(Node,Holdout_LASSO,by.x="Gene",by.y="TG",all.x=T,all.y=F)

# Get releavant information on the nodes. IE is this a placenta speific gene
Node$Gene<-as.character(Node$Gene)

Node$PlacentaGene<-rep("NO",dim(Node)[1])
Node$PlacentaGene[match(HPA_Genes,Node$Gene)]<-"YES"
table(Node$PlacentaGene)


```

#Calculate Network Centrality
```{r}
net <- graph_from_data_frame(d=Edge[,3:4],vertices=Node, directed=T)

##C Part 2 Calculate Network Centrality Scores ####

# 1. Degree Centrality: The number of links held by each node. A local
#measure which does NOT take into account network size. IN OUR TRN, Degree
#Centrality would reflect the number of target genes for each TF

Degree_Centrality <- centr_degree(net, mode = "all")$res

Network_Scores<-cbind(Node,Degree_Centrality)
rownames(Network_Scores)<-Network_Scores$Gene

#This is a sanity check: This should be the same as the # of Nodes


#closeness centrality: a global score that reveals how close each node is to all
#other nodes based on shortest path length

Network_Scores$Closeness_Centrality<-closeness(net, mode="all")


#Betweenness Centrality: The number of shortest paths going through a vertex or edge

Network_Scores$Betweeness_Centrality<-estimate_betweenness(net, vids = V(net), directed = TRUE,cutoff=0,
                                                           weights = NULL, nobigint = TRUE)

#Eigenvector Centrality: In graph theory, eigenvector centrality (also called eigencentrality) is a
#measure of the influence of a node in a network. Relative scores are assigned to all nodes in the
#network based on the concept that connections to high-scoring nodes contribute more to the score of
#the node in question than equal connections to low-scoring nodes. A high eigenvector score means
#that a node is connected to many nodes who themselves have high scores.

Network_Scores$Eigenvector_Centrality<- evcent(net)$vector

##Kleinberg's hub centrality scores.
#https://igraph.org/r/doc/hub_score.html


```

# Make Figure of Network Centrality Scores and TFs 
```{r}

#TF Plot
Plot_Mat<-Network_Scores[order(Network_Scores$Degree_Centrality,decreasing=T),]
PlotTFs<-subset(Plot_Mat,Type=="TF")
HubTFs<-subset(PlotTFs,Degree_Centrality>500) #110 Hub
TopTFs<-PlotTFs[1:15,]

CheckTFs<-PlotTFs[16:404,]
CheckTFs<-subset(CheckTFs,Eigenvector_Centrality>0.4)

Colors<-brewer.pal(2,"Dark2")
par(mfrow=c(1,1))
plot(PlotTFs$Degree_Centrality,PlotTFs$Eigenvector_Centrality,pch=20,cex=0.5,xlab="Genes Sorted by Degree Centrality",ylab="Eigen Vector Centrality",xlim=c(0,2500))
points(HubTFs$Degree_Centrality,HubTFs$Eigenvector_Centrality,pch=20,cex=0.75,col=Colors[2])
points(TopTFs$Degree_Centrality,TopTFs$Eigenvector_Centrality,pch=20,cex=1.1,col=Colors[1])
abline(v=500,lwd=1.25,lty=2,col=Colors[2])
text(TopTFs$Degree_Centrality,TopTFs$Eigenvector_Centrality,labels=TopTFs$Gene,cex=0.5,pos=4,col=Colors[1])
text(CheckTFs$Degree_Centrality,CheckTFs$Eigenvector_Centrality,labels=CheckTFs$Gene,cex=0.5,pos=2,col=Colors[2])


TFNetworkResults<-PlotTFs[,c(1,6:10)]
```


Examine Nodes Up and Downsteram
```{r}
TFMat<-subset(Node,Type=="TF")

TFs_Downstream<-rep(NA,dim(TFMat)[1])

for(i in 1:dim(TFMat)[1]){
  temp1<- subset(Edge, TF == TFMat$Gene[i]) #Pull out those which it is a regulator
  TFs_Downstream[i]<-length(intersect(temp1$targetGene,TFMat$Gene))
}


TFs_Upstream<-rep(NA,dim(TFMat)[1])

for(i in 1:dim(TFMat)[1]){
  temp1<- subset(Edge, targetGene == TFMat$Gene[i]) #Pull out those which it is a regulator
  TFs_Upstream[i]<-length(intersect(temp1$TF,TFMat$Gene))
}

TFMat$OutOfDegree<-TFs_Downstream
TFMat$InDegree<-TFs_Upstream

```

#Key Information for Paper
```{r}
table(TG_Table$RSquared>0.5)
median(TF_Table$N_Targets)
table(TF_Table$N_Targets>500)
x<-subset(Node,Type=="TF")
x<-subset(x,PlacentaGene=="YES")

```

# Save Key Files
```{r}
#R File is for when We prune for Preterm Birth Genes
save(Node,Edge,file="~/PlacentalTRN/FinalResults/Node&Edge.RData")

#.csv Files are for loading into Cytoscape
write.csv(Node,file="~/PlacentalTRN/FinalResults/NodeFile.csv")
write.csv(Edge,file="~/PlacentalTRN/FinalResults/EdgeFile.csv")

#Write the Model For Ryan to Use
Model$Q<-p.adjust(Model$P,method="BH",n=length(Model$P))
write.csv(Model,file="~/PlacentalTRN/FinalResults/TRNForRShinyApp.csv")
write.csv(Holdout_LASSO,file="~/PlacentalTRN/FinalResults/TargetGeneRSquared.csv")


#Write the Supplmental Table with Network Connectivity Scores
write.csv(TFNetworkResults,file="~/PlacentalTRN/FinalResults/SuppTableNetworkResults.csv")

#Might as well also write the supplemental table with the interactions
write.csv(Interactions_Sig,file="~/PlacentalTRN/FinalResults/SignificantInteractions.csv")
```
