---
title: "AnalyzingSexInteractionsTRN"
author: "Alison Paquette"
date: "9/29/2020"
output: html_document
---
```{r}
load("~/PlacentalTRN/IntermediateData/FetalSexInteractionTerms.RData")
load("/local/users/apaquett/PlacentalTRN/PreprocessedData/Covar_All_3182019.RData")
load("~/PlacentalTRN/IntermediateData/ProcessedRNASequencingData6192020.RData")
```

# Once Again Stratify Gene Expression matrixes
```{r}
Covar_all<-Covar_all[colnames(Exprs_all),]
colnames(Exprs_all)==rownames(Covar_all)

###Stratify Expression Matrix
Male_Matrix<-as.data.frame(t(Exprs_all[,rownames(subset(Covar_all,Fetal_Sex=="Male"))]))
Female_Matrix<-as.data.frame(t(Exprs_all[,rownames(subset(Covar_all,Fetal_Sex=="Female"))]))
Matrix<-as.data.frame(t(Exprs_all))
rownames(Matrix)==rownames(Covar_all)


Interactions_Sig<-subset(Interactions,Q_interaction<0.05)
rownames(Interactions_Sig)<-paste0(Interactions_Sig$Target_Gene,"_",Interactions_Sig$TF)


```

Perform Assesment of how Gene interactions are working
-STronger positive or negative interactions in females vs. males
-or
-stronger positive or negative assocations in males vs. females
```{r}
Stronger_Males<-subset(Interactions_Sig,Diff_MvF>0)
table(Stronger_Males$Estimate_Males>0)
table(Stronger_Males$Estimate_Females>0)

Stronger_Females<-subset(Interactions_Sig,Diff_MvF<0)
table(Stronger_Females$Estimate_Males>0)
table(Stronger_Females$Estimate_Females>0)

## Q1: Stronger Positive Values in Males####
Stronger_MalesPos<-subset(Stronger_Males,Estimate_All>0)
Stronger_MalesPos<-Stronger_MalesPos[order(Stronger_MalesPos$Diff_MvF,decreasing=T),]

#Q2: Positive Diff, but negative in males, means the interction is actually stronger in females (Significantly less negative)###
Stronger_MalesNeg<-subset(Stronger_Males,Estimate_All<0)
Stronger_MalesNeg<-Stronger_MalesNeg[order(Stronger_MalesNeg$Diff_MvF,decreasing=T),]


##Q3: Stronger In Females
Stronger_FemalesPos<-subset(Stronger_Females,Estimate_All>0)
Stronger_FemalesPos<-Stronger_FemalesPos[order(Stronger_FemalesPos$Diff_MvF,decreasing=F),]

## Q4: Stronger in females; negative. stronger and more negative in females
Stronger_FemalesNeg<-subset(Stronger_Females,Estimate_All<0)
Stronger_FemalesNeg<-Stronger_FemalesNeg[order(Stronger_FemalesNeg$Diff_MvF,decreasing=F),]

library(RColorBrewer)
colors_TRN<-brewer.pal(n = 8, name = "Dark2")

par(mfrow=c(1,1))
plot(Interactions_Sig$Estimate_All,Interactions_Sig$Diff_MvF,pch=16,cex=0.5,xlab="Estimate: TF/Target Gene",ylab="Difference: Males vs. Females",xlim=c(-3,3),ylim=c(-3,3))

#Quadrant1
points(Stronger_MalesPos$Estimate_All,Stronger_MalesPos$Diff_MvF,pch=16,cex=0.75,col=colors_TRN[3])
text(Stronger_MalesPos$Estimate_All[1:3],Stronger_MalesPos$Diff_MvF[1:3],labels=rownames(Stronger_MalesPos)[1:3],cex=0.75,pos=4,col=colors_TRN[3])

# Qudrant 2
points(Stronger_FemalesPos$Estimate_All,Stronger_FemalesPos$Diff_MvF,pch=16,cex=0.75,col=colors_TRN[1])
text(Stronger_FemalesPos$Estimate_All[1:3],Stronger_FemalesPos$Diff_MvF[1:3],labels=rownames(Stronger_FemalesPos)[1:3],cex=0.75,pos=4,col=colors_TRN[1])

#Qudarant 3
points(Stronger_MalesNeg$Estimate_All,Stronger_MalesNeg$Diff_MvF,pch=16,cex=0.75,col=colors_TRN[4])
text(Stronger_MalesNeg$Estimate_All[1:3],Stronger_MalesNeg$Diff_MvF[1:3],labels=rownames(Stronger_MalesNeg)[1:3],cex=0.75,pos=4,col=colors_TRN[4])


# Qudrant 4
points(Stronger_FemalesNeg$Estimate_All,Stronger_FemalesNeg$Diff_MvF,pch=16,cex=0.75,col=colors_TRN[2])
text(Stronger_FemalesNeg$Estimate_All[1:3],Stronger_FemalesNeg$Diff_MvF[1:3],labels=rownames(Stronger_FemalesNeg)[1:3],cex=0.75,pos=2,col=colors_TRN[2])


abline(v=0,lwd=1.25,lty=2,col="azure4")
abline(h=0,lwd=1.25,lty=2,col="azure4")

```

# Plot the top interactions within each "quadrant" so that we can visually see what they look like
```{r}
Plot_Interact<-function(Matrix,Male_Matrix,Female_Matrix,TF,Target.Gene){
  
  
  TF_All<-Matrix[,TF]
  Target.Gene_All<-Matrix[,Target.Gene]
  
  TF_M<-Male_Matrix[,TF]
  Target.Gene_M<-Male_Matrix[,Target.Gene]
  
  TF_F<-Female_Matrix[,TF]
  Target.Gene_F<-Female_Matrix[,Target.Gene]
  
  
  
  plot(TF_All,Target.Gene_All,pch=16,xlab=print(TF),ylab=print(Target.Gene))

  #Males
  points(TF_M,Target.Gene_M,pch=16,col="blue")
  abline(lm(Target.Gene_M~TF_M),lty=2,col="blue",lwd=1.5)
  
  
  points(TF_F,Target.Gene_F,pch=16,col="deeppink2")
  abline(lm(Target.Gene_F~TF_F),lty=2,col="deeppink2",lwd=1.5)
  
}

par(mfrow=c(1,3))
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_MalesPos$TF[1],Stronger_MalesPos$Target_Gene[1]) # Lower in Females than Males
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_MalesPos$TF[2],Stronger_MalesPos$Target_Gene[2]) # Lower in Females than Males
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_MalesPos$TF[3],Stronger_MalesPos$Target_Gene[3]) # Lower in Females than Males

### Qudrant 2
par(mfrow=c(1,3))
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_MalesNeg$TF[1],Stronger_MalesNeg$Target_Gene[1]) # Lower in Females than Males
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_MalesNeg$TF[2],Stronger_MalesNeg$Target_Gene[2]) # Lower in Females than Males
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_MalesNeg$TF[3],Stronger_MalesNeg$Target_Gene[3]) # Lower in Females than Males


### Qudrant 3
par(mfrow=c(1,3))
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_FemalesPos$TF[1],Stronger_FemalesPos$Target_Gene[1]) # Lower in Females than Males
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_FemalesPos$TF[2],Stronger_FemalesPos$Target_Gene[2]) # Lower in Females than Males
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_FemalesPos$TF[3],Stronger_FemalesPos$Target_Gene[3]) # Lower in Females than Males

### Qudrant 4
par(mfrow=c(1,3))
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_FemalesNeg$TF[1],Stronger_FemalesNeg$Target_Gene[1]) # Lower in Females than Males
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_FemalesNeg$TF[2],Stronger_FemalesNeg$Target_Gene[2]) # Lower in Females than Males
Plot_Interact(Matrix,Male_Matrix,Female_Matrix,Stronger_FemalesNeg$TF[3],Stronger_FemalesNeg$Target_Gene[3]) # Lower in Females than Males


```

###---Map TFs & target gene to genomic regions
```{r}
library(biomaRt)
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))


TFTable<-as.data.frame(table(Interactions_Sig$TF))
colnames(TFTable)<-c("Gene","NumberInteractions")
TFTable$Type<-"TF"
rownames(TFTable)<-TFTable$Gene

G_list <- getBM(filters= "hgnc_symbol", attributes= c("hgnc_symbol","chromosome_name","start_position"),values=(TFTable$Gene),mart= mart)

G_list<-subset(G_list,chromosome_name=="1"|chromosome_name=="2"|chromosome_name=="3"|chromosome_name=="4"|chromosome_name=="5"|
chromosome_name=="6"|chromosome_name=="7"|chromosome_name=="8"|chromosome_name=="9"|chromosome_name=="10"|chromosome_name=="11"|
chromosome_name=="12"|chromosome_name=="13"|chromosome_name=="14"|chromosome_name=="15"|chromosome_name=="16"|chromosome_name=="17"|
chromosome_name=="18"|chromosome_name=="19"|chromosome_name=="20"|chromosome_name=="21"|chromosome_name=="22"|chromosome_name=="X"|chromosome_name=="Y")

rownames(G_list)<-G_list$hgnc_symbol

TFTable<-merge(TFTable,G_list,by='row.names',all=F)




TGTable<-as.data.frame(table(Interactions_Sig$Target_Gene))
colnames(TGTable)<-c("Gene","NumberInteractions")
TGTable$Type<-"TargetGene"
rownames(TGTable)<-TGTable$Gene

check <- getBM(filters= "hgnc_symbol", attributes= c("hgnc_symbol","chromosome_name","start_position"),values=(colnames(Matrix)),mart= mart)


G_list <- getBM(filters= "hgnc_symbol", attributes= c("hgnc_symbol","chromosome_name","start_position"),values=(TGTable$Gene),mart= mart)

G_list<-subset(G_list,chromosome_name=="1"|chromosome_name=="2"|chromosome_name=="3"|chromosome_name=="4"|chromosome_name=="5"|
                 chromosome_name=="6"|chromosome_name=="7"|chromosome_name=="8"|chromosome_name=="9"|chromosome_name=="10"|chromosome_name=="11"|
                 chromosome_name=="12"|chromosome_name=="13"|chromosome_name=="14"|chromosome_name=="15"|chromosome_name=="16"|chromosome_name=="17"|
                 chromosome_name=="18"|chromosome_name=="19"|chromosome_name=="20"|chromosome_name=="21"|chromosome_name=="22"|chromosome_name=="X"|chromosome_name=="Y")

rownames(G_list)<-G_list$hgnc_symbol

TGTable<-merge(TGTable,G_list,by='row.names',all=F)

All<-rbind(TFTable,TGTable)

table(TFTable$chromosome_name)
table(TGTable$chromosome_name)

x<-as.data.frame(table(All$chromosome_name))
hold<-as.factor(1:22)
x$Var1 <- ordered(c("1", "2", "3", "4", "5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","X"))

barplot(x$Freq,ylim=c(0,300),xlab="Chromosome",ylab="Number of Interactions",names.arg=x$Var1)
```
